refactor(transformer, minifier): replace `AstBuilder::move_xxxx` methods with `TakeIn` trait

Dunqing opened the pull request:
Pure refactor, `TakeIn` trait is exactly used for these cases, we don't need an indirect call for `move_xxx`.

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10170?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#10171** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10171?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#10170** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10170?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10170?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#10169** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10169?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=10170).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented:
> I did this in #9970 but then decided against it, for the reasons given in #10001. I think `take_in` makes it too easy to pull large types onto the stack, when you could likely find a way to use `move_expression` or `move_statement` instead (which are much cheaper).
> 
> I'm not dead set on that position if you strongly disagree, but that's my overall feeling.

I think we should have one way of doing things. Making decisions on `TakeIn` or `move_xxx` is really confusing ...

Dunqing committed:
refactor(transformer, minifier): replace `AstBuilder::move_xxxx` methods with `TakeIn` trait (#10170)

Pure refactor, `TakeIn` trait is exactly used for these cases, we don't need an indirect call for `move_xxx`.

overlookmotel commented:
@Dunqing Should we add an `AstBuilder::take` method, so you don't have to use `TakeIn` explicitly?

```diff
- declaration.take_in(self.ast.allocator)
+ self.ast.take(declaration)
```

I don't mind either way, just asking what you think.

Dunqing commented:
> @Dunqing Should we add an `AstBuilder::take` method, so you don't have to use `TakeIn` explicitly?
> 
> ```diff
> - declaration.take_in(self.ast.allocator)
> + self.ast.take(declaration)
> ```
> 
> I don't mind either way, just asking what you think.

Yes, this would be easier to use, I don't have opinion on this.

overlookmotel commented:
OK, I opened an issue: #10314
